### DISCUSSION TITLE
Welcome to phase4 Discussions!

### DISCUSSION THREAD

**Discussion by phax** [phax/phase4#45] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jsalichos** [comment]
Hello,

I'm John from Argus Software. I'm exploring the idea of creating an access point service in Greece but I'm relatively new to PEPPOL idea, so, I'll need help. I like the AS4 project because it is written in .NET/C# and has an interesting architecture.

I don't really get the details of the concept yet but I'm trying.

Am I in the right place?

Best regards,

John

**Reply by phax** [reply]
Hi John,
I think I need to disappoint you here - this project is also written in Java and not .NET.
I am only aware of one Open Source solution of AS4 in .net and that one is called As4.net maintained by the EC. But I am not aware how Peppol specific this solution is.
hth

**Comment by shruti9899** [comment]
Hey,

what is the basic differences between peppol, cef and like other profiles? What are the factors to consider to choose either of them?
I am noob here ;) Help

Thanks,
Shruti

**Reply by phax** [reply]
Hi Shruti,
good question, but hard to answer :)

When you are looking at it purely from an AS4-perspective, than the "Peppol AS4 profile 2.x" is based on the "CEF AS4 profile 1.14". Therefore Peppol further limits the CEF profile. But CEF also mandates the usages of at least TLS 1.2 with certain cipher suites, and that was inherited by Peppol as well. With phase4 you have predefined configurations for Peppol, CEF and ENTSOG, and if you implement it in a smart way, you can even handle them all in parallel.

But if you also look at the "dynamic discovery" aspects of Peppol (SMP/SML) there are some bigger differences, as Peppol uses different SMP and SML specifications (see https://docs.peppol.eu/edelivery/) compared to CEF which is relying on the OASIS versions of the standards (see http://docs.oasis-open.org/bdxr/bdx-smp/v1.0/bdx-smp-v1.0.html and http://docs.oasis-open.org/bdxr/BDX-Location/v1.0/BDX-Location-v1.0.html). A comparison of the SMP standards can be found at https://github.com/phax/phoss-smp/wiki/Difference-between-Peppol-and-OASIS-BDXR

The factors for decision making very much depend on your goal :) To participate in Peppol you need to support the Peppol AS4 profile - no choice here. If you however plan to participate in other EU projects, that also deal with eDelivery, than you will most likely need the CEF/OASIS specifications. If you plan to set something up for internal use, I (personally) would start from the CEF solution and further limit it to what I need. But since there are already a lot of CEF compliant solutions in the market, I think it is safe to consider the CEF AS4 profile a stable foundation for your next steps.

hth, Philip

**Reply by shruti9899** [reply]
cool. Thanks. Its helpful.
I want to know is there any step by step documentation of how to use CEF client, profile and how to configure phase4 library to run? That would be helpful for further steps.

**Reply by phax** [reply]
Well that very much depends on your use case - what do you want to achieve. Do you need this "dynamic discovery" stuff or not? Do you want to send and/or receive? What application server are you using etc. etc.

Sending is definivitely easier, as you can see from one of the example clients: https://github.com/phax/phase4/blob/master/phase4-cef-client/src/test/java/com/helger/phase4/cef/MainPhase4CEFSenderToop.java - you just configure what you want to send and the rest is magically done for you.

When receiving, it is a bit more tricky - I recommend you have a look at https://github.com/phax/phase4/tree/master/phase4-server-webapp for an example application that just receives and than discards the incoming message.

**Comment by albert-patron** [comment]
Hello,

I'am a java developper and I need to implement the Peppol Access Point AS4 to send invoices to clients. 

I'am trying to implement one of your subproject phase4-peppol-client solutions but I don't understand how to set some of the parameters of the Phase4PeppolSender.builder() and how to set the properties file parameters.
I've already the PKI v3 AP Test certificate.
Can anyone help me?
Thanks in advance

Eco

**Reply by phax** [reply]
Hi Eco,
I tried to give you a quick introduction:
* You need a "Document Type ID" and "Process ID" - based on these identifiers the payload needs to be defined. For allowed values, please see the code lists at https://docs.peppol.eu/edelivery/codelists/ - to send a "Default Peppol Invoice", the Document Type Identifier is `busdox-docid-qns::urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1` and the Process Identifier is `cenbii-procid-ubl::urn:fdc:peppol.eu:2017:poacc:billing:01:1.0`
* For the sender participant Identifier, you can initially choose whatever you like. Whoever it needs to follow the structure `[0-9]{4}:.+` (as in `9915:phase4-test-sender`). The first 4 digits are called "Participant Identifier Scheme" and also need to follow the respective code list at https://docs.peppol.eu/edelivery/codelists/
* The receiver participant Identifier needs to be structured in the same way. For testing purposes you can try to send to `9915:helger`
* The sender Party ID is the `CN` part of the "Subject" field in your Peppol AP certificate (follows the RegEx `P[A-Z]{2}\d{6}`)
* The payload is (e.g.) the Invoice that you want to send. You can provide it as a byte array or as a pre-parse XML DOM element. Please make sure that it matches the validation rules of this particular form (as provided by the respective Schematron rules) - see also the `validationConfiguration` parameters in case you want to use the built-in validations.
* With the SMP client you effectively chose to use Dynamic Discovery (that would go to far as an intro) and select, whether you chose the Peppol production network (`ESML.DIGIT_PROD`) or the Peppol test network (`ESML.DIGIT_TEST`)
* All the other parameters and callbacks depend on your use case.

To enable the dumping of incoming and outgoing messages, add this to your code at the beginning:
```java
   AS4DumpManager.setIncomingDumper (new AS4IncomingDumperFileBased ());
   AS4DumpManager.setOutgoingDumper (new AS4OutgoingDumperFileBased ());
```
Hope that is a good starting point.

Best, Philip

**Reply by albert-patron** [reply]
Hi Philip, 
Thanks a lot for your fast and clear answer, 

For the next step, I need to create a wss4j apache server, correct?
Once again, thanks a lot for your help,

Eco

**Reply by phax** [reply]
Hi Eco,
* So for sending, you don't need a server application but a standalone "main" is sufficient. So it should be easy to embed it into another application
* For receiving messages, you need a "Server" component, that is waiting for requests to come in. This can be imho done with the Spring-Boot application (But I am not an expert in this).
* With the submodule `phase4-peppol-server-webapp` I am providing an alternative approach - a web application (WAR) that can be deployed in an existing Tomcat or Jetty. The server itself is the Tomcat or Jetty - it routes the requests to the correct application
* You don't need to deal with "WSS4J" - this is all well-hidden in phase4. You effectively "only" need to accept a request from the outside and then pass it into phase4. Using the Servlet technology is easiest for me. But other people are also managed to use phase4 from within AWS Lambdas or from other application frameworks.
* In the end, the exchange of AS4 is "just" an exchange of specifically formatted messages over HTTP.
hth, Philip

**Reply by ashoktronix** [reply]
I have mailed you personally from my official email as well, Any steps you have how to use for ENTSOG profile has a receiver ?

**Comment by albert-patron** [comment]
Hi Philip,

Thanks for your help,
I've created a standalone "main", but I'am getting errors,
I'am thinking this is related with my phase4.properties file configuration.

![errorCompile](https://user-images.githubusercontent.com/101553656/158540890-6bfaf94b-a27b-4949-bc00-ccd4e96ddb83.PNG)

My phase4.properties:
![phase4PropertiesConfig](https://user-images.githubusercontent.com/101553656/158541947-4ba98e0b-df63-45e8-a8da-b87d02a850ca.PNG)



Can anyone help?
Thanks in advance

**Reply by phax** [reply]
Hi Eco,
I think it is very simple: it seems like your `phase4.properties` file is not found.
I recommend to place all settings in `application.properties` to avoid confusion. Because the order of configuration item resolution can be quite difficult, and e.g. if the same property is used in `application.properties` and `phase4.properties`, the one in `phase4.properties`  is used.

Where are the configuration files searched?
* `phase4.properties` is searched in the current work directory (depends on the application server) or in the classpath (e.g. in the `WEB-INF/classes` folder)
* `application.properties` is searched in the same places as `phase4.properties`, but alternatively you can e.g. specify the Java System Property `config.file` (or the ENV variable `CONFIG_FILE`) to point to an absolute path where this file is searched. The full resolution logic is described at https://github.com/phax/ph-commons/wiki/ph-config

Therfore placing all properties in `application.properties` is preferred. Than define a system property or Env Variable to point to that file. And inside the configuration file I recommend to only use absolute paths (like `/path/to/teste.p7b`) for the keystore. For the trust store you can leave it as it is, because it is resolved from the ClassPath.

**Comment by albert-patron** [comment]
Hi Philip,
Now I've another error:
![errorCompile](https://user-images.githubusercontent.com/101553656/158587215-b851bdbb-da07-4b95-930f-ac192af96e25.PNG)

My new phase4.properties:
![phase4PropertiesConfig1](https://user-images.githubusercontent.com/101553656/158587960-938fb735-1b01-4c5e-804f-dd24cae1af6f.PNG)

Thanks in advance for your help

**Reply by phax** [reply]
Well, that still means, that the correct "phase4.properties" file is not yet used. Did you place it in the right place? Feel free to email me to avoid disclosing too many details (see the pom.xml for my email address).
